### PR TITLE
Ope 1757

### PR DESCRIPTION
**Pull request by palegre-fluendo**
- Add tag (test id) to the failed test output
- Do not consider all scenarios failed if some in a .feature crashes or deadlocks.
- Fix deadlock after test timeout

### CODE REVIEW

File: behave/reporter/summary.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def print_failing_scenarios(self, stream=None):
 
         stream.write("\nFailing scenarios:\n")
         for scenario in self.failed_scenarios:
-            stream.write(u"  %s  %s\n" % (scenario.location, scenario.name))
+            stream.write(u"  %s %s  %s\n" % (', '.join(t for t in scenario.tags if len(t.split('-'))==2 ), scenario.location, scenario.name))
```
Comment (JohnJudd) [inline behave/reporter/summary.py:165]: Please comment code to make it more readable

File: behave/runner.py [hunk 2]
```diff
@@ -31,6 +32,18 @@
 else:
     import traceback
 
+import time
+
+class Chrono:
```
Comment (JohnJudd) [inline behave/runner.py]: Chrono maybe a bit cryptic or general IMO i think time TimeoutTimer would be a clearer name

File: behave/runner.py [hunk 3]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
 
         foreachitem(feature)
 
-    def _run_feature(self, feature, q):
-        try:
-            ret = feature.run(self)
+    class QueueEvent(Enum):
+        ITEM_STARTED = 0
```
Comment (JohnJudd) [inline behave/runner.py]: It would be clearer to be called scenario, as it each item is actually a scenario

File: behave/runner.py [hunk 4]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
...
+        ITEM_FINISHED = 1
+        FEATURE_FINISHED = 2
```
Comment (JohnJudd) [inline behave/runner.py:668]: Featured finished means the feature file so the *.feature has completed, maybe some comments explaining what exactly they represent.

File: behave/runner.py [hunk 5]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
...
 
+    def _run_feature(self, feature, q, starting):
```
Comment (JohnJudd) [inline behave/runner.py]: queue would be clearer than just q.

File: behave/runner.py [hunk 6]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
...
+        try:
             def clearTraceback(item, param):
                 if hasattr(item, 'exc_traceback'):
                     delattr(item, 'exc_traceback')
 
+            def before_run_item(run_item):
+                if starting > run_item.test_id:
```
Comment (JohnJudd) [inline behave/runner.py]: starting_test_id would be a clearer name.

File: behave/runner.py [hunk 7]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
...
+                    return True
+                if not hasattr(run_item, 'started'):
+                    setattr(run_item, 'started', True)
+                q.put((ModelRunner.QueueEvent.ITEM_STARTED, run_item))
+                return False
+
+            def after_run_item(run_item):
+                # replace tags array with new strings to avoid <class 'super'> after memory release
```
Comment (JohnJudd) [inline behave/runner.py]: Some more comments about the reason for this, pickling etc

File: behave/runner.py [hunk 8]
```diff
@@ -644,21 +657,42 @@ def foreachitem(item):
...
+                run_item.tags = [str(tag) for tag in run_item.tags]
+
+                self._forEachItemInFeatures(run_item, clearTraceback, 0)
+                q.put((ModelRunner.QueueEvent.ITEM_FINISHED, run_item))
+                return False
```
Comment (JohnJudd) [inline behave/runner.py]: No need to return as no one is expecting a return value

File: behave/runner.py [hunk 9]
```diff
@@ -689,37 +723,63 @@ def read_timeout(tags):
...
+                                error = StatusError.timeout
+                                break
+
+                        if not proc.is_alive():
+                            if proc.exitcode < 0:
+                                print("Child Process finished with exit code {}".format(proc.exitcode), flush=True)
+                                error = StatusError.crash
+                                break
+
+                    if error != StatusError.none:
```
Comment (JohnJudd) [inline behave/runner.py:771]: A few lines explaining the code and the reson for it will be great :)

File: behave/runner.py (hunk 5 above)
Comment (palegre-fluendo) [reply behave/runner.py]: I will rename it to queue

File: behave/runner.py (hunk 4 above)
Comment (palegre-fluendo) [reply behave/runner.py:668]: added comments for all of them

File: behave/runner.py (hunk 6 above)
Comment (palegre-fluendo) [reply behave/runner.py]: will change to 'next', i think it is clear

File: behave/runner.py (hunk 9 above)
Comment (palegre-fluendo) [reply behave/runner.py:771]: great

**Review by JohnJudd — APPROVED** [review]

**Pull request merged**